### DISCUSSION TITLE
Could Lowdefy be deployed with a private cloud?

### DISCUSSION THREAD

**Discussion by xiexj2** [lowdefy/lowdefy#668] [Questions and Help] [answered]
Hi,

The idea of Lowdefy is amazing for the low-code development by yaml. According to the documents, Lowdefy can be deployed with AWS Lambda, Docker and Netlify. Could it be deploy with a private cloud, which doesn't connect to Internet? Are there any guides for this?

Thanks in advance

**Comment by SamTolmay** [comment, marked as answer]
Glad you are enjoying building apps with Lowdefy.

If you are able to deploy Docker applications (If you have a Kubernetes cluster for example), you should be able to deploy Lowdefy using a Docker image. Unfortunately, we don't have any examples or guides on this.

You will need to host your own blocks inside your private cloud. The Lowdefy blocks are served from a CDN, at https://blocks-cdn.lowdefy.com/v${version}. You will need to build the blocks packages, and host the `dist/` artefacts somewhere in your private cloud on a file server. For example, to host the Ant Design blocks we serve:
https://blocks-cdn.lowdefy.com/v3.18.0/blocks-antd/meta/Button.json
https://blocks-cdn.lowdefy.com/v3.18.0/blocks-antd/remoteEntry.js


Then, in your `lowdefy.yaml` file, you will need to specify URLs for the blocks you use, this would be something like:
```
types:
  Button:
    url: {your_domain}/meta/Button.json
 TextInput:
    url: {your_domain}/meta/TextInput.json
```

Making hosting your own blocks like this easier is something we are considering adding as an enterprise feature.

Please feel free to get in touch with us at hello [at] lowdefy.com, and we can help you set this up.

**Reply by xiexj2** [reply]
Thanks for your response! We'll follow your advices.

**Reply by SamTolmay** [reply]
It's a pleasure! You can find the blocks packages at https://github.com/lowdefy/lowdefy/tree/main/packages/blocks. If you run `yarn build` in the root of the project repository that will build all the packages, including the blocks.

**Reply by xiexj2** [reply]
Your suggestion is very effective, and now I can deploy the lowdefy app and its blocks in our private cloud. Thanks again.
However, I still feel a little troublesome if every URL of the blocks has to be specified in the `lowdefy.yaml` file. Is there a earier way? For example, could I modify or config the `renderer` or `servers` module to directly point to the URL of local Blocks?

**Reply by SamTolmay** [reply]
That's great!

The default locations are specified in the following file:
https://github.com/lowdefy/lowdefy/blob/main/packages/build/src/utils/meta/defaultMetaLocations.js

We could make the blocks server URL (currently hard-coded to `https://blocks-cdn.lowdefy.com`) a CLI parameter, so when you run the build command you run:
```
lowdefy build --blocks-server <your-blocks-url>
```

Would that work for you? 

The best solution I can think of for the current version is to write a script that generates the block URLs as JSON and then referencing that JSON file in your apps. You could then use that file across apps as well. So that would look like:

```yaml
# lowdefy.yaml
lowdefy: 3.18.0

types:
  _ref: blocks.json
```
with
```json
// blocks.json
{
  "Button": {
      "url": "{your_domain}/meta/Button.json"
  }
} 
```

**Reply by xiexj2** [reply]
The solution with `blocks.json` can make the lowdefy app run correctly, too.  At the same time I am looking forward to the method of parameter configuration, which will be more elegant. Thanks.

**Reply by SamTolmay** [reply]
I think that will add a lot of value, I have created a feature request issue, #670, so you can track this there.

**Reply by SamTolmay** [reply]
In v3.20 we have added a build configuration variable that allows you to specify the URL from which blocks are served, as discussed above.

This can either be set using the `--blocks-server-url` command-line option, or in the `lowdefy.yaml` file as:
```yaml
lowdefy: 3.21.0

cli:
  blocksServerUrl: https://example.com
```
The build outputs from the different blocks packages then need to be served from that URL, with each package in at a path corresponding to the package name (`@lowdefy/blocks-basic` should be served from `https://example.com/blocks-basic/` ).

You can find more details [here](https://docs.lowdefy.com/cli)

**Reply by xiexj2** [reply]
Thanks. I would like to ask further, how to start the blocks server？

**Reply by SamTolmay** [reply]
The blocks server needs to be a static file server, that serves the build output for each blocks package (under a path corresponding to the package name).

We use an AWS S3 bucket configured as a static website, and use Cloudflare as a DNS and CDN. You can see the Github action we use to build the assets and upload them to S3 here: https://github.com/lowdefy/lowdefy/blob/main/.github/workflows/release.yaml

If you are looking at hosting the blocks yourself, something like NGINX would be needed.

**Reply by xiexj2** [reply]
Thanks Sam. I have two more questions.

1. The renderer component also uses blocks-cdn.lowdefy.com to lazy-loading, and it is written to lowdefy/lowdefy-main/packages/cli/src/commands/dev/shell/bootstrap.js. If we also want to be loaded from the local, whether to modify the source code?
2. In order to `yarn start` from blocksAntd for block server testing, we change the localhost to the local cloud server address in the file `lowdefy/packages/blocks/blocksAntd/webpack.demo.js` , but for the browser this change will take effect with a delay of about 1 hour for unknown reasons, and it feels that it is cached by the lowdefy app. After checking, it is found that the cache is in localowdefy/.lowdefy/.cache/scripts/build/3_21_2/package/dist, and it will not be cleaned automatically every time. After modifying the configuration, we need to manually clear the scripts in this directory, otherwise the browser will continue to GET  `Http://localhost:${port}/remontEntry.js` before modification.

```
//lowdefy/packages/blocks/blocksAntd/webpack.demo.js
const addRemoteEntryUrl = (content, absoluteFrom) => {
  const scope = sanitizeName(packageJson.name);
  const meta = JSON.parse(content);
  meta.moduleFederation = {
    module: path.basename(absoluteFrom, '.json'),
    scope,
    version: packageJson.version,
    //remoteEntryUrl: `http://localhost:${port}/remoteEntry.js`,
    remoteEntryUrl: `http://*.*.*.*:${port}/remoteEntry.js`,
  };
  return JSON.stringify(meta);
};

```

**Reply by SamTolmay** [reply]
Hi @xiexj2 

Regarding 1:
I forgot about the renderer also being pulled in from the CDN when answering this question, sorry about that. There are two cases where we need to worry about the renderer, the first being in the development server in the CLI, and the second being in production servers.

For the development server, we could modify the code ar `/packages/cli/src/commands/dev/shell/bootstrap.js` to respect the `--blocks-server-url` option, so it will then also read from the server you are hosting. This isn't something that is to hard to do, and we should be able to include it in the next release.

For production servers, the renderer does not need to be pulled in dynamically like the development server. I think it would be best if we modify them to be part of the production server bundle, as this should also make the production servers slightly faster and more robust. This is something that is a little more complicated than the development server, so I can't commit to a timeline for this.

To track this I have opened the following issues: #840, #841 


Regarding 2:
I am not exactly sure what your problem is here.

The CLI does cache both the build and server scripts, as well as blocks metadata in the `.lowdefy/.cache` directory. This can be cleaned by running `lowdefy clean-cache`, or by just deleting the directory. You should not need to delete the scripts, since they will be fetched from npm again and remain the same. The blocks metadata cache is cleaned everytime the `dev` command is run. So if you do change the metadata being served from your blocks server, the cache should be cleaned, or the server should be restarted and this is supposed to clean the cache. You can also check that the module federation metadata in the JSON files created by build (found at `.lowdefy/build`) are correct. Hope that answers the question.

**Reply by xiexj2** [reply]
Thanks SamTolmay, I'll check it.

**Comment by GanizaniSitara** [comment]
I couldn't comment on https://github.com/lowdefy/lowdefy/issues/928 anymore, but the solution for me was to get Symantec to classify the URL (blocks-cdn.lowdefy.com), which was unclassified by them and set to blocking on corporate proxy. The link for symantec is here https://sitereview.bluecoat.com/#/ Just wanted to let you know.

**Reply by Gervwyk** [reply]
Great! Thanks, @GanizaniSitara for letting us know. For the next major version of Lowdefy (v4) all apps will run their own build bundle which will include all blocks etc, so the external cdn will not be required. Also, apps will be so much faster 🚀

**Comment by mingfang** [comment]
I think it would be a huge improvement if we didn't have to hardcode the remoteEntryUrl into the meta json file.
The build tool can be enhancement to detect a relative remoteEntryUrl and then construct the full url using the same host as the meta url.
For example. If we have a custom type like this

```yaml
name: my-app
lowdefy: 3.10.1
types:
  DisplayBlock:
    url: http://dev:3002/meta/DisplayBlock.json
```

and the DisplayBlock.json has remoteEntryUrl as `remoteEntryUrl: /remoteEntry.js`.

Then the build tool should automatically change the remoteEntryUrl to be `remoteEntryUrl: http://dev:3002/remoteEntry.js`.  It looks like that can be done here https://github.com/lowdefy/lowdefy/blob/07286dfa8471e6ed293c33c7b37ab59003ccb5c6/packages/build/src/utils/meta/getMeta.js#L54

The advantage with this would be I can deploy my block server anywhere without having to rebuild to change the meta json file.
